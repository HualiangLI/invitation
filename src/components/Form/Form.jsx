/**
 * Created by 李华良 on 2019/4/7
 */
import React from 'react'
import PropTypes from 'prop-types'
import './Form.scss'

class Form extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      name: '',
      email: '',
      confirmEmail: '',
    }
  }

  handleFullNameChange = e => {
    this.setState({ name: e.target.value })
  }
  handleEmailChange = e => {
    this.setState({ email: e.target.value })
  }
  handleConfirmEmailChange = e => {
    this.setState({ confirmEmail: e.target.value })
  }

  handleFormSubmit = e => {
    e.preventDefault()
    const {
      onSubmit,
    } = this.props
    const {
      name,
      email,
      confirmEmail,
    } = this.state
    onSubmit({
      name,
      email,
      confirmEmail,
    })
  }

  render() {
    const {
      name,
      email,
      confirmEmail,
    } = this.state

    const {
      errMsg,
      loading,
    } = this.props

    const isBtnDisabled = loading

    return (
      <form className="comp__form" onSubmit={this.handleFormSubmit}>
        <div className="form-item">
          <input type="text" placeholder="Full name" value={name} onChange={this.handleFullNameChange}/>
        </div>
        <div className="form-item">
          <input type="text" placeholder="Email" value={email} onChange={this.handleEmailChange}/>
        </div>
        <div className="form-item">
          <input type="text" placeholder="Confirm email" value={confirmEmail} onChange={this.handleConfirmEmailChange}/>
        </div>
        <div className="form-btn-row">
          <button type="submit" disabled={isBtnDisabled}>{ loading ? 'Sending, please wait...' : 'Send' }</button>
        </div>
        {errMsg && (
          <p className="form-errs">{ errMsg }</p>
        )}
      </form>
    )
  }
}

Form.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  errMsg: PropTypes.string.isRequired,
  loading: PropTypes.bool.isRequired,
}

export default Form
