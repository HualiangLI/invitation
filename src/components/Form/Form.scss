/**
 * Created by 李华良 on 2019/4/7
 */
.comp__form {
  input {
    width: 100%;
    padding: 5px 5px;
    border: 1px solid #ccc;

    &:focus {
      border-color: #333333;
    }
  }

  .form-item {
    font-size: 0;
    margin-bottom: 10px;
  }
  .form-btn-row {
    padding-top: 10px;

    button {
      text-align: center;
      width: 100%;
      line-height: 2;
    }
  }
  .form-errs {
    text-align: center;
    color: #e57373;
    line-height: 1.5;
    padding-top: 10px;
  }
}
