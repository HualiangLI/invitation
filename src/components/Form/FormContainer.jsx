/**
 * Created by 李华良 on 2019/4/7
 */
import React from 'react'
import PropTypes from 'prop-types'
import axios from 'axios'

import Form from './Form'

class FormContainer extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      errMsg: '',
      loading: false,
    }
  }

  handleSubmit = ({ name, email, confirmEmail }) => {
    let errMsg = ''
    if (!name) errMsg = 'Please input your full name.'
    else if (!email) errMsg = 'Please input your email.'
    else if (!/^([A-Za-z0-9_\-\.])+\@([A-Za-z0-9_\-\.])+\.([A-Za-z]{2,4})$/.test(email)) errMsg = 'Unrecognized email.'
    else if (!confirmEmail) errMsg = 'Please confirm your inputted email.'
    else if (email !== confirmEmail) errMsg = 'Email doesn\'t match.'
    if (errMsg) {
      this.setState({ errMsg })
      return Promise.resolve()
    }
    this.requestInvitation({ name, email })
  }

  requestInvitation(data) {
    this.setState({ loading: true })
    return axios.post(
      'https://l94wc2001h.execute-api.ap-southeast-2.amazonaws.com/prod/fake-auth',
      data,
      {
        headers: {
          'Content-Type': 'application/json',
        },
      }
    )
      .then(() => new Promise((resolve, reject) => {
          this.setState({ loading: false }, resolve)
        }))
      .then(this.props.onSuccess)
      .catch(err => {
        let errMsg = 'Request failed.'
        if (err.response && err.response.data) errMsg = err.response.data.errorMessage
        else if (err.message) errMsg = err.message
        this.setState({ errMsg, loading: false })
      })
  }

  render() {
    const {
      errMsg,
      loading
    } = this.state
    return (
      <Form onSubmit={this.handleSubmit} errMsg={errMsg} loading={loading}/>
    )
  }
}

FormContainer.propTypes = {
  onSuccess: PropTypes.func.isRequired
}

export default FormContainer
