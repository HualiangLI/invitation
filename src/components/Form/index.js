/**
 * Created by 李华良 on 2019/4/7
 */
import Form from './FormContainer'

export default Form
