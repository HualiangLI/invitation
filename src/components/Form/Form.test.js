import React from 'react'
import { mount, shallow } from 'enzyme'
import Form from './Form'

describe('Form', () => {
  it('should render correctly', () => {
    const wrapper = mount(<Form onSubmit={jest.fn()} errMsg='you have errors' loading={false} />)
    expect(wrapper.render()).toMatchSnapshot()
  })

  it('should not call onSubmit when loading', function () {
    const onSubmit = jest.fn()
    const wrapper = shallow(<Form onSubmit={onSubmit} errMsg='you have errors' loading={true} />)
    wrapper.find('button').simulate('click')
    expect(onSubmit).not.toBeCalled()
  })

  it('should submit object with full name, email and confirmEmail', function () {
    const onSubmit = jest.fn()
    const state = { name: 'test name', email: 'test email', confirmEmail: 'test confirmEmail' }
    const wrapper = shallow(<Form onSubmit={onSubmit} errMsg='' loading={false} />)
    wrapper.setState(state)
    wrapper.find('form').simulate('submit', { preventDefault: () => {} })
    expect(onSubmit).toBeCalledWith(expect.objectContaining(state))
  })
})
