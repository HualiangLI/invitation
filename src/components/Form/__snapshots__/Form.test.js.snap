// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "comp__form",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "form-item",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "placeholder": "Full name",
              "type": "text",
              "value": "",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "form-item",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "placeholder": "Email",
                "type": "text",
                "value": "",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "form-item",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "placeholder": "Confirm email",
                  "type": "text",
                  "value": "",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "placeholder": undefined,
                  "type": undefined,
                  "value": undefined,
                },
                "x-attribsPrefix": Object {
                  "placeholder": undefined,
                  "type": undefined,
                  "value": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "form-btn-row",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "type": "submit",
                  },
                  "children": Array [
                    Object {
                      "data": "Send",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "text",
                    },
                  ],
                  "name": "button",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "type": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "type": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "form-errs",
                },
                "children": Array [
                  Object {
                    "data": "you have errors",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "p",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "form-item",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "placeholder": "Email",
              "type": "text",
              "value": "",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "form-item",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "placeholder": "Confirm email",
                "type": "text",
                "value": "",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "form-btn-row",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "type": "submit",
                },
                "children": Array [
                  Object {
                    "data": "Send",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "button",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "type": undefined,
                },
                "x-attribsPrefix": Object {
                  "type": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "form-errs",
              },
              "children": Array [
                Object {
                  "data": "you have errors",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "p",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "form-item",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "placeholder": "Full name",
                "type": "text",
                "value": "",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "form-item",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "placeholder": "Confirm email",
              "type": "text",
              "value": "",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "placeholder": undefined,
              "type": undefined,
              "value": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "form-btn-row",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "type": "submit",
              },
              "children": Array [
                Object {
                  "data": "Send",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "form-errs",
            },
            "children": Array [
              Object {
                "data": "you have errors",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "p",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "form-item",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "placeholder": "Email",
                "type": "text",
                "value": "",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "form-item",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "placeholder": "Full name",
                  "type": "text",
                  "value": "",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "placeholder": undefined,
                  "type": undefined,
                  "value": undefined,
                },
                "x-attribsPrefix": Object {
                  "placeholder": undefined,
                  "type": undefined,
                  "value": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "form-btn-row",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "type": "submit",
            },
            "children": Array [
              Object {
                "data": "Send",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "button",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "type": undefined,
            },
            "x-attribsPrefix": Object {
              "type": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "form-errs",
          },
          "children": Array [
            Object {
              "data": "you have errors",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "p",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "form-item",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "placeholder": "Confirm email",
                "type": "text",
                "value": "",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "placeholder": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "form-item",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "placeholder": "Email",
                  "type": "text",
                  "value": "",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "placeholder": undefined,
                  "type": undefined,
                  "value": undefined,
                },
                "x-attribsPrefix": Object {
                  "placeholder": undefined,
                  "type": undefined,
                  "value": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "form-item",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "placeholder": "Full name",
                    "type": "text",
                    "value": "",
                  },
                  "children": Array [],
                  "name": "input",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "placeholder": undefined,
                    "type": undefined,
                    "value": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "placeholder": undefined,
                    "type": undefined,
                    "value": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "form-errs",
        },
        "children": Array [
          Object {
            "data": "you have errors",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "p",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "form-btn-row",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "type": "submit",
              },
              "children": Array [
                Object {
                  "data": "Send",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "form-item",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "placeholder": "Confirm email",
                  "type": "text",
                  "value": "",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "placeholder": undefined,
                  "type": undefined,
                  "value": undefined,
                },
                "x-attribsPrefix": Object {
                  "placeholder": undefined,
                  "type": undefined,
                  "value": undefined,
                },
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "form-item",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "placeholder": "Email",
                    "type": "text",
                    "value": "",
                  },
                  "children": Array [],
                  "name": "input",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "placeholder": undefined,
                    "type": undefined,
                    "value": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "placeholder": undefined,
                    "type": undefined,
                    "value": undefined,
                  },
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "form-item",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "placeholder": "Full name",
                      "type": "text",
                      "value": "",
                    },
                    "children": Array [],
                    "name": "input",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "placeholder": undefined,
                      "type": undefined,
                      "value": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "placeholder": undefined,
                      "type": undefined,
                      "value": undefined,
                    },
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "form",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
